### DISCUSSION TITLE
Segmentation loss computation

### DISCUSSION THREAD

**Discussion by someuser2026** [ultralytics/ultralytics#22239] [Q&A]
Hi, a very basic question regarding loss for segmentation network. I can see it is a combination of DFL, Box loss and Mask loss, however,

1. What are the exact losses used?
2. How does the library process the .txt polygon annotations for the loss? Does it convert it into segmentation mask on the fly and then computes the losses?

Thank you

**Comment by UltralyticsAssistant** [comment]
👋 Hello @someuser2026, thanks for starting a discussion about segmentation loss in Ultralytics 🚀. This is an automated response to help you get timely guidance while an Ultralytics engineer reviews your post and will assist you soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including:
- Exact Ultralytics version and environment details
- Model and task used (e.g., YOLO11-seg)
- A small sample of your .txt polygon annotations
- Relevant training logs and any code/config snippets
- Confirmation you’re following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/)

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Reply by glenn-jocher** [reply]
Great questions—here’s how Ultralytics YOLO handles seg training:

1) Exact losses: segmentation uses a 4-term loss vector [box, seg, cls, dfl]. Box = CIoU on positive anchors, plus Distribution Focal Loss (DFL) on distances; Cls = BCEWithLogits against soft target scores; Seg = BCEWithLogits between predicted masks and GT masks, cropped to the target box and normalized by object area. See the implementations in [v8SegmentationLoss](https://docs.ultralytics.com/reference/utils/loss/#ultralyticsutilslossv8segmentationloss), [BboxLoss (CIoU + DFL)](https://docs.ultralytics.com/reference/utils/loss/#ultralyticsutilslossbboxloss), and the mask path via `single_mask_loss`/`calculate_segmentation_loss` on the same page.

2) Polygon .txt processing: yes—polygons from YOLO-seg labels are rasterized to binary masks during dataloading (not stored as images). This uses utilities like [polygon2mask / polygons2masks / polygons2masks_overlap](https://docs.ultralytics.com/reference/data/utils/#ultralyticsdatautilspolygon2mask), then masks are downsampled to the prototype resolution and consumed in `calculate_segmentation_loss`, where predictions are formed as `pred_mask = torch.einsum('in,nhw->ihw', pred, proto)`.
